private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+		gitDefaultRemoteURL = strings.TrimSuffix(gitDefaultRemoteURL, ".git")
+		// Then we get the current HEAD commit.
+		currentHEADCommit, err := getCurrentHEADGitCommit(ctx, runner, input)
+		if err != nil {
+			return nil, err
+		}
+		// Bitbucket is the only URL that uses the "/commits" route, both Github and Gitlab
+		// use "/commit", so we default to that if the remote URL does not point at Bitbucket.
+		if strings.HasPrefix(gitDefaultRemoteURL, bitBucketURL) {
+			return bufmodule.UploadWithSourceControlURL(fmt.Sprintf(
+				bitBucketRemoteURLFormat,
+				gitDefaultRemoteURL,
+				currentHEADCommit,
+			)), nil
+		}
+		return bufmodule.UploadWithSourceControlURL(fmt.Sprintf(
+			defaultGitRemoteURLFormat,
+			gitDefaultRemoteURL,
+			currentHEADCommit,
+		)), nil
+	}
+	// Otherwise we sort by alphabetical order and return the first URL
+	sortedRemotes := slicesext.MapKeysToSortedSlice(remoteToURL)
+	if len(sortedRemotes) > 0 {
+		return bufmodule.UploadWithSourceControlURL(remoteToURL[sortedRemotes[0]]), nil
+	}
+	return nil, nil
+}
+
+func getCurrentHEADGitCommit(
+	ctx context.Context,
+	runner command.Runner,
+	input string,
+) (string, error) {
+	buffer := bytes.NewBuffer(nil)
+	if err := runner.Run(
+		ctx,
+		gitCommand,
+		command.RunWithArgs("rev-parse", "HEAD"),
+		command.RunWithStdout(buffer),
+		command.RunWithStderr(buffer),
+		command.RunWithDir(input),
+	); err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(buffer.String()), nil
+}
+
+func getGitDefaultBranch(
+	ctx context.Context,
+	runner command.Runner,
+	envContainer app.EnvContainer,
+	remotes []string,
+	input string,
+) (string, error) {
+	remoteToHEADBranch := map[string]string{}
+	for _, remote := range remotes {
+		buffer := bytes.NewBuffer(nil)
+		if err := runner.Run(
+			ctx,
+			gitCommand,
+			command.RunWithArgs("remote", "show", remote),
+			command.RunWithStdout(buffer),
+			command.RunWithStderr(buffer),
+			command.RunWithDir(input),
+			command.RunWithEnv(app.EnvironMap(envContainer)),
+		); err != nil {
+			return "", err
+		}
+		branch, err := getHEADBranchFromGitRemoteOutput(buffer.Bytes())
+		if err != nil {
+			return "", err
+		}
+		remoteToHEADBranch[remote] = branch
+	}
+	// We prioritze the Git default remote, "origin", HEAD branch as the default branch
+	if gitDefaultRemoteHEADBranch, ok := remoteToHEADBranch[gitDefaultRemote]; ok {
+		return gitDefaultRemoteHEADBranch, nil
+	}
+	// Otherwise we sort by alphabetical order and return the first URL
+	sortedRemotes := slicesext.MapKeysToSortedSlice(remoteToHEADBranch)
+	if len(sortedRemotes) > 0 {
+		return remoteToHEADBranch[sortedRemotes[0]], nil
+	}
+	return "", nil
+}
+
+// This checks for the HEAD branch from the output of `git remote show <remote>`.
+// sed -n '/HEAD branch/s/.*: //p'
+func getHEADBranchFromGitRemoteOutput(output []byte) (string, error) {
+	// First find where "HEAD branch"
+	index := bytes.Index(output, []byte("HEAD branch: "))
```

Comment:
Instead of dealing in bytes, I think this would by much easier to read/maintain with a bufio.Scanner, strings.Cut, etc.

Reply:
Ah yes, it's a lot better that way!